### ISSUE TITLE
[STORAGE-S3] Add Support for Custom S3 Endpoint

### ISSUE THREAD

**Issue by Sebastien73** [airbytehq/airbyte#44174]
### Helm Chart Version

0.429.0

### What step the error happened?

Upgrading the Platform or Helm Chart

### Description

Currently, the log4j2-s3.xml configuration file in the Airbyte platform is hardcoded to work with Amazon S3 as the default storage option. However, when deploying Airbyte via the Helm chart, it's not possible to override the default AWS endpoint (amazonaws.com) to use S3-compatible storage services from other providers (e.g., Scaleway).

To illustrate, it's not possible to set an endpoint like https://s3.fr-par.scw.cloud for Scaleway’s S3-compatible storage service.

For instance, it is not currently possible to use an endpoint like `https://s3.fr-par.scw.cloud`.

### Proposal

To enhance compatibility with other S3-compatible storage services, I propose adding support for a custom S3 endpoint in the log4j2-s3.xml file. Below is a suggested change:

```xml
<Properties>
    <Property name="ci-mode">${sys:ciMode:-false}</Property>
    <!--
    date format is datadog friendly so that it can try to detect multilines logs
    https://github.com/DataDog/datadog-agent/blob/a27c16c05da0cf7b09d5a5075ca568fdae1b4ee0/pkg/logs/internal/decoder/auto_multiline_handler.go#L208

    trace_id and span_id are added to the logs so that datadog can correlate logs and traces https://docs.datadoghq.com/tracing/other_telemetry/connect_logs_and_traces/
    https://app.datadoghq.com/logs/pipelines?search=env%3Aci
     -->
    <Property name="pattern-with-trace-id">%d{yyyy-MM-dd HH:mm:ss,SSS}{GMT+0} [dd.trace_id=%X{dd.trace_id} dd.span_id=%X{dd.span_id}] %p %C{1.}(%M):%L %replace{%m}{apikey=[\w\-]*}{apikey=*****}%n</Property>
    <!-- Mask the string apikey=<string> to apikey=***** to prevent secrets leaking. -->
    <Property name="default-pattern">%d{yyyy-MM-dd HH:mm:ss}{GMT+0} %highlight{%p} %C{1.}(%M):%L - %replace{%m}{apikey=[\w\-]*}{apikey=*****}%n</Property>
    <!--Logs the timestamp and log_source/application name in the beginning of the line if it exists with a > separator, and then always the rest of the line.-->
    <Property name="simple-pattern">%d{yyyy-MM-dd HH:mm:ss}{GMT+0}%replace{ %X{log_source}}{^ -}{} > %replace{%m}{apikey=[\w\-]*}{apikey=*****}%n</Property>

    <!-- Always log INFO by default. -->
    <Property name="log-level">${sys:LOG_LEVEL:-${env:LOG_LEVEL:-INFO}}</Property>

    <Property name="route-ttl">${env:LOG_IDLE_ROUTE_TTL:-15}</Property>

    <!-- Note that logging to S3 will leverage the DefaultAWSCredentialsProviderChain for auth. -->
    <Property name="s3-bucket">${sys:STORAGE_BUCKET_LOG:-${env:STORAGE_BUCKET_LOG:-}}</Property>
    <Property name="s3-region">${sys:AWS_DEFAULT_REGION:-${env:AWS_DEFAULT_REGION:-}}</Property>
    <!-- Add the following property to allow custom S3 endpoint -->
    <Property name="s3-endpoint">${sys:AWS_ENDPOINT_URL:-${env:AWS_ENDPOINT_URL:-}}</Property>
</Properties>
```
In addition to modifying the log4j2-s3.xml file, a corresponding change would be necessary in the `values.yaml` file within the `Helm chart` to support custom S3 endpoints. The change should be made in the global.storage section, as shown below:

```yaml
storage:
  # -- The storage backend type. Supports s3, gcs, minio (default)
  type: s3 # change to your preferred storage type
  # -- Secret name where storage provider credentials are stored
  #storageSecretName: "airbyte-config-secrets"

  # S3
  bucket: ## S3 bucket names that you've created. We recommend storing the following all in one bucket.
    log: airbyte-bucket
    state: airbyte-bucket
    workloadOutput: airbyte-bucket
  s3:
    region: "" ## e.g. us-east-1
    authenticationType: credentials ## Use "credentials" or "instanceProfile"

#############################################
# Add the following to support custom S3 endpoint:
    endpoint: <CUSTOM_ENDPOINT_URL>
#############################################

  # GCS
  #bucket: ## GCS bucket names that you've created. We recommend storing the following all in one bucket.
  #  log: airbyte-bucket
  #  state: airbyte-bucket
  #  workloadOutput: airbyte-bucket
  #gcs:
  #  projectId: <project-id>
  #  credentialsJson: /secrets/gcs-log-creds/gcp.json

```

#### Expected Outcome

These changes will allow users to specify an alternative S3-compatible endpoint when configuring Airbyte with non-AWS S3 storage providers. This enhancement will improve the flexibility and compatibility of Airbyte, enabling its deployment in a broader range of cloud storage environments.

**Comment by aqeelat** [comment]
try setting
```yaml
global:
  storage:
    minio:
      endpoint: "your custom s3 endpoint"
```

**Comment by Sebastien73** [comment]
> try setting
> 
> ```yaml
> global:
>   storage:
>     minio:
>       endpoint: "your custom s3 endpoint"
> ```

I have tested this, but with "minio" at the storage type, I can't change the default value of the region.
My S3 use "fr-par" region.

With "minio" storage type, I can change the endpoint but not the region and when I use "s3" storage type, I can change the region but not the endpoint.

**Comment by djpirra** [comment]
So storage type should be set as S3 and have a minio entry under storage?

If storage type is set to minio it will create a minio instance.

**Comment by Sebastien73** [comment]
Hello, 

When I used the S3 storage type in my values.yaml, I can't set a custom endpoint. This value stay by default at aws endpoint.
This is why I tested the use of the "minio" storage type, because this type has the possibility of having a customizable endpoint except the region parameter which remains by default on AWS.

I would like to have possibilities to custom "region" and  "endpoint" parameter for used object storage with Scaleway provider.

First test - With s3 storage type : 

```
 globale:
   storage:
       type: s3
         bucket: ## S3 bucket names that you've created. We recommend storing the following all in one bucket.
         activityPayload: airbyte-s3
         log: airbyte-s3
         state: airbyte-s3
         workloadOutput: airbyte-s3
      s3:
        region: fr-par ## e.g. us-east-1
        authenticationType: credentials ## Use "credentials" or "instanceProfile"
        accessKeyIdSecretKey: AWS_ACCESS_KEY_ID
        secretAccessKeySecretKey: AWS_SECRET_ACCESS_KEY
        
        It's here where I need to have something to custom endpoint like the "region" parameter.
```

Second test - With minio storage type : 

```
 globale:
   storage:
       type: minio
          endpoint : custom endpoint

      ANd here I can't change the value of the "region" parameter. 
```

**Comment by ssaunier** [comment]
Being able to use S3-compatible services of different IaaS providers (Digital Ocean, Scaleway, etc.) and not just AWS S3 would be a great addition indeed 👍 !

**Comment by marcosmarxm** [comment]
@airbytehq/platform-deployments and @perangel sharing this issue as important to the community open source users.

**Comment by andrewtran1314** [comment]
any updates on this new feature?

**Comment by nfla** [comment]
I need it too ;). Waiting for update.  
Thank you.

**Comment by chris922** [comment]
I am a bit confused if s3-compatible storage should already work or not. I have the same issues mentioned in this thread using an own hosted MinIO instance. I am not able to configure the endpoint via Helm chart.
Nevertheless, within the docu there is something mentioned that S3 compatible storage should work: https://docs.airbyte.com/platform/deploying-airbyte/integrations/storage

> Currently, S3, GCS, and Azure are supported. If you are using an S3 compatible solution, use the S3 type and provide an endpoint key/value as needed.

I tried setting a `endpoint` key + value within my S3 secret, but it doesn't work. Any ideas?

**Comment by NerushKirill** [comment]
Hi everyone! 
We are really looking forward to this opportunity! 
We have an s3 compatible DigitalOcean that we would like to use.

Thanks!

**Comment by alyapin** [comment]
Hi everyone,
Any updates on the issue? We’re also waiting on support for the custom endpoint.

**Comment by ticup** [comment]
Same here. Docs suggests it works with S3 compatible solutions, but can't find a way to provide the endpoint, so it tries to connect to aws. (Trying to get it to work with Digital Ocean object storage).

**Comment by ticup** [comment]
I eventually got it to work for Digital Ocean as follows...

values.yaml:

````
global:
  storage:
    type: "minio"
    minio:
      endpoint: <S3 endpoint, including region>
    secretName: airbyte-config-secrets # Name of your Kubernetes secret.
    bucket: ## S3 bucket names that you've created. We recommend storing the following all in one bucket.
      log: airbyte-prod
      state: airbyte-prod
      workloadOutput: airbyte-prod
````

airbyte-config-secrets.yaml

````
apiVersion: v1
kind: Secret
metadata:
  name: airbyte-config-secrets
  namespace: airbyte
type: Opaque
stringData:
  MINIO_ACCESS_KEY_ID: <S3 access key id>
  MINIO_SECRET_ACCESS_KEY: <S3 secret access key>
  minio-access-key-id: <S3 access key id>
  minio-secret-access-key: <S3 secret access key>
  s3-access-key-id: <S3 access key id>
  s3-secret-access-key: <S3 secret access key>
````

**Comment by hasanozciftcii** [comment]
> I eventually got it to work for Digital Ocean as follows...
> 
> values.yaml:
> 
> ```
> global:
>   storage:
>     type: "minio"
>     minio:
>       endpoint: <S3 endpoint, including region>
>     secretName: airbyte-config-secrets # Name of your Kubernetes secret.
>     bucket: ## S3 bucket names that you've created. We recommend storing the following all in one bucket.
>       log: airbyte-prod
>       state: airbyte-prod
>       workloadOutput: airbyte-prod
> ```
> 
> airbyte-config-secrets.yaml
> 
> ```
> apiVersion: v1
> kind: Secret
> metadata:
>   name: airbyte-config-secrets
>   namespace: airbyte
> type: Opaque
> stringData:
>   MINIO_ACCESS_KEY_ID: <S3 access key id>
>   MINIO_SECRET_ACCESS_KEY: <S3 secret access key>
>   minio-access-key-id: <S3 access key id>
>   minio-secret-access-key: <S3 secret access key>
>   s3-access-key-id: <S3 access key id>
>   s3-secret-access-key: <S3 secret access key>
> ```

Hi @ticup, it also worked on my end. Thank you for sharing!

**Comment by sunst0rm** [comment]
@ticup  @hasanozciftcii  

Can you help please?

I am also trying to dep[loy Airbyte using custom S3 (we are using Ceph) and it does not work:

```
global:
  storage:
    type: "minio"
    minio:
      endpoint: "https://s3.company.com"
    secretName: airbyte-config-secrets
    bucket:
      log: airbyte
      state: airbyte
      workloadOutput: airbyte
```

I also added:

```
minio:
    enabled: false
````


My secret is:
```
---

apiVersion: v1
kind: Secret
metadata:
  name: airbyte-config-secrets
type: Opaque
stringData:
  MINIO_ACCESS_KEY_ID: "{{ airbyte_s3_bucket_access_key }}"
  MINIO_SECRET_ACCESS_KEY: "{{ airbyte_s3_bucket_secret_key }}"
  minio-access-key-id: "{{ airbyte_s3_bucket_access_key }}"
  minio-secret-access-key: "{{ airbyte_s3_bucket_secret_key }}"
  s3-access-key-id: "{{ airbyte_s3_bucket_access_key }}"
  s3-secret-access-key: "{{ airbyte_s3_bucket_secret_key }}"
```

Can you help what could be the issue?

Helm gets crazy and does not want to deploy it and also copy values from airbyte-config-secrets to airbyte-airbyte-secrets

Could you post your whole values if possible?

**Comment by hasanozciftcii** [comment]
> [@ticup](https://github.com/ticup) [@hasanozciftcii](https://github.com/hasanozciftcii)
> 
> Can you help please?
> 
> I am also trying to dep[loy Airbyte using custom S3 (we are using Ceph) and it does not work:
> 
> ```
> global:
>   storage:
>     type: "minio"
>     minio:
>       endpoint: "https://s3.company.com"
>     secretName: airbyte-config-secrets
>     bucket:
>       log: airbyte
>       state: airbyte
>       workloadOutput: airbyte
> ```
> 
> I also added:
> 
> ```
> minio:
>     enabled: false
> ```
> 
> My secret is:
> 
> ```
> ---
> 
> apiVersion: v1
> kind: Secret
> metadata:
>   name: airbyte-config-secrets
> type: Opaque
> stringData:
>   MINIO_ACCESS_KEY_ID: "{{ airbyte_s3_bucket_access_key }}"
>   MINIO_SECRET_ACCESS_KEY: "{{ airbyte_s3_bucket_secret_key }}"
>   minio-access-key-id: "{{ airbyte_s3_bucket_access_key }}"
>   minio-secret-access-key: "{{ airbyte_s3_bucket_secret_key }}"
>   s3-access-key-id: "{{ airbyte_s3_bucket_access_key }}"
>   s3-secret-access-key: "{{ airbyte_s3_bucket_secret_key }}"
> ```
> 
> Can you help what could be the issue?
> 
> Helm gets crazy and does not want to deploy it and also copy values from airbyte-config-secrets to airbyte-airbyte-secrets
> 
> Could you post your whole values if possible?

my values like below and it is currently working. You can create a new secret with different name

```
storage:
    type: ""
    secretName: airbyte-secret-dev
    bucket:
      log: "dev-airbyte-bucket"
      state: "dev-airbyte-bucket"
      workloadOutput: "dev-airbyte-bucket"
    minio:
      endpoint: "http://minio.<namespace-name>.svc.cluster.local:9000"
      region: "us-east-1"
```

```
apiVersion: v1
kind: Secret
metadata:
  name: airbyte-secret-dev
type: Opaque
stringData:
  MINIO_ACCESS_KEY_ID: ".."
  MINIO_SECRET_ACCESS_KEY: ".."
  minio-access-key-id: ".."
  minio-secret-access-key: ".."
  s3-access-key-id: ".."
  s3-secret-access-key: ".."
```

**Comment by sunst0rm** [comment]
I wouldn't think that leaving empty value in type will make it work. Thanks @hasanozciftcii 

This is my working config if anyone encounters same problem:
```
minio:
  enabled: false

global:
  storage:
    type: ""
    secretName: "airbyte-config-secrets"
    minio:
      endpoint: "https://s3.example.com"
      region: "default"
    bucket:
      log: airbyte
      state: airbyte
      workloadOutput: airbyte
```

and secrets:

```
apiVersion: v1
kind: Secret
metadata:
  name: airbyte-config-secrets
type: Opaque
stringData:
  MINIO_ACCESS_KEY_ID: "{{ airbyte_s3_bucket_access_key }}"
  MINIO_SECRET_ACCESS_KEY: "{{ airbyte_s3_bucket_secret_key }}"
  minio-access-key-id: "{{ airbyte_s3_bucket_access_key }}"
  minio-secret-access-key: "{{ airbyte_s3_bucket_secret_key }}"
  s3-access-key-id: "{{ airbyte_s3_bucket_access_key }}"
  s3-secret-access-key: "{{ airbyte_s3_bucket_secret_key }}"
```

**Comment by Sebastien73** [comment]
I had forgotten to follow this issue to see any updates ^^

I deployed Airbyte 1.8.2 with Helm chart v1 today, and I can confirm that your solution is still the best way to use an S3-compatible storage.

@ticup thanks for your answer, that help me !

I’m not sure if Helm chart v2 resolves this, or if we still need to use your solution.

**Comment by AliJakobs** [comment]
are there any updates? @sunst0rm s solution doesnt work for me.

**Comment by vitas-pm** [comment]
My current Chart version: 2.1.0

After wasting almost an entire day yesterday trying to get the log storage to work with our third-party object storage provider I have found the following issues in the airbyte-platform repository:

Your storage config forces a choice between s3 and minio as the STORAGE_TYPE, but neither option exposes both an endpoint override and a region at the same time. Not via Helm values, not via env vars, not via any documented config path. This makes it impossible to point Airbyte at most third-party S3-compatible object stores (Backblaze B2, Cloudflare R2, DigitalOcean Spaces, Wasabi, on-prem MinIO clusters that require a specific region, etc.) without patching source.

### Root cause
1. **S3StorageConfig has a region field but no endpoint field, and its client builder never calls endpointOverride:**

`airbyte-commons-micronaut/src/main/kotlin/io/airbyte/micronaut/runtime/AirbyteConfig.kt:1010-1018`

```
@ConfigurationProperties("s3")
@Requires(property = STORAGE_TYPE, pattern = "(?i)^s3$")
data class S3StorageConfig(
  val accessKey: String = "",
  val secretAccessKey: String = "",
  val region: String = "",
) : StorageConfig
```

`airbyte-commons-storage/src/main/kotlin/io/airbyte/commons/storage/StorageClient.kt:572-580`

```
internal fun AirbyteStorageConfig.S3StorageConfig.s3Client(): S3Client {
  val builder = S3Client.builder().region(Region.of(this.region))
  if (this.accessKey.isNotBlank()) {
    builder.credentialsProvider { AwsBasicCredentials.create(this.accessKey, this.secretAccessKey) }
  }
  return builder.build()
}
```

No .endpointOverride(...) is ever invoked for this path, it can only ever talk to real AWS S3.

2. **MinioStorageConfig has an endpoint field but no region field, and the region is hardcoded:**

`AirbyteConfig.kt:996-1005`

```
@ConfigurationProperties("minio")
@Requires(property = STORAGE_TYPE, pattern = "(?i)^minio$")
data class MinioStorageConfig(
  val accessKey: String = "",
  val secretAccessKey: String = "",
  val endpoint: String = "",
) : StorageConfig
```

`StorageClient.kt:558-566`

```
internal fun AirbyteStorageConfig.MinioStorageConfig.s3Client(): S3Client =
  S3Client.builder()
    .serviceConfiguration { it.pathStyleAccessEnabled(true) }
    .credentialsProvider { AwsBasicCredentials.create(this@s3Client.accessKey, this@s3Client.secretAccessKey) }
    .endpointOverride(URI(this@s3Client.endpoint))
    // The region isn't actually used but is required. Set to us-east-1 based on https://github.com/minio/minio/discussions/15063.
    .region(Region.US_EAST_1)
```

The region is unconditionally us-east-1 regardless of what the target provider actually requires.

3. **The Helm chart's generated env mapping has the identical gap, confirming this isn't just a doc gap but a config-schema gap:**

`charts/v2/airbyte/config.yaml:1067-1071`: `s3.region` → `AWS_DEFAULT_REGION`, gated to `discriminatorOpts: [s3]`. No `s3.endpoint` key exists anywhere in the schema.

`charts/v2/airbyte/config.yaml:1128-1132`: `minio.endpoint` → `MINIO_ENDPOINT`, gated to `discriminatorOpts: [minio]`. No `minio.region` key exists anywhere in the schema.

Since the discriminator (`storage.type`) forces exactly one of s3/minio, there is no combination of Helm values that produces both MINIO_ENDPOINT/equivalent and a non-default region.

4. **Bonus dead-config bug found in the same area:** 
The Helm chart also exposes `storage.minio.s3PathStyleAccess` → `S3_PATH_STYLE_ACCESS` (`config.yaml:1134-1138`), but the application never reads that env var anywhere, `StorageClient.kt:561` hardcodes pathStyleAccessEnabled(true) unconditionally. Setting `storage.minio.s3PathStyleAccess: false` in `values.yaml` has silently no effect.

### Expected behavior
Either:

`S3StorageConfig` gains an optional `endpoint` field that's passed to `.endpointOverride(...)` when set, or
`MinioStorageConfig` gains a `region` field that's passed to `.region(...)` instead of the hardcoded `Region.US_EAST_1`, or
The two configs are merged into one "S3-compatible" config with both `endpoint` (optional) and `region` fields.
Any of these should be threaded through `AirbyteConfig.kt`, `StorageClient.kt`, and the Helm `config.yaml`/`values.yaml` schema.

Affected files:
`airbyte-commons-micronaut/src/main/kotlin/io/airbyte/micronaut/runtime/AirbyteConfig.kt` (lines 996-1018)
`airbyte-commons-storage/src/main/kotlin/io/airbyte/commons/storage/StorageClient.kt` (lines 558-580)
`charts/v2/airbyte/config.yaml` (lines 1067-1071, 1128-1138)
`charts/v2/airbyte/values.yaml` (lines 187-221)


Endpoint override alongside region is about as basic as S3-compatible support gets. That this issue is still open and ignored for almost two years for a simple config option that is the standard in any helm chart working with S3 and even after the v2 rework is very disappointing and in my opinion even somewhat embarrassing. Especially for a product whose core self-hosted pitch is infrastructure flexibility and data sovereignty. Your S3 source and destination connectors are able to do that, so why not your internal program core?

Now, because of that oversight I had to concede and continue using the Minio option. But theres another, very similar issue here. The storage config for the PVC is missing the `storageClass` option. 

The volumeClaimTemplates block in `charts/v2/airbyte/templates/minio.yaml:18-25` only exposes `accessModes` and `resources.requests.storage`:

```
volumeClaimTemplates:
    - metadata:
        name: airbyte-minio-pv-claim
      spec:
        accessModes: [ "ReadWriteOnce" ]
        resources:
          requests:
            storage: {{ .Values.minio.storage.volumeClaimValue }}
```

Again, an option that is almost as basic as the storage size itself, when it comes to working with PVCs. So I had to do, and I'm not kidding, this:

```
minio:
  storage:
    volumeClaimValue: "5Gi\n          storageClassName: csi-disk-topology"
```

Which is more of an exploit but at least it's working, unlike the other issue which is hardcoded into the program itself.

It really doesn't seem like that much effort is needed to fix these issues. But if none of you want to tackle it, I'd be willing to open a PR to finally close this issue.
